File: C/2399-check-distances-between-same-letters/main.go
package main

func checkDistances(s string, distance []int) bool {
	valid := make([]int, 26)
	pos := make([]int, 26)
	for i, b := range s {
		offset := int(byte(b) - 'a')
		valid[offset]++
		if valid[offset] == 1 {
			pos[offset] = i
		} else {
			pos[offset] = i - pos[offset] - 1
		}
	}
	// fmt.Println(valid, pos)
	for i := 0; i < 26; i++ {
		if valid[i] == 2 {
			if pos[i] != distance[i] {
				return false
			}
		}
	}
	return true
}

File: C/2399-check-distances-between-same-letters/main.rs
struct Solution {}

impl Solution {
  pub fn check_distances(s: String, distance: Vec<i32>) -> bool {
    let mut buf: [i32; 26] = [-1; 26];
    s.as_bytes()
      .iter()
      .enumerate()
      .map(|(idx, b)| {
        let offset = (b - 'a' as u8) as usize;
        if buf[offset] > -1 && idx as i32 - buf[offset] - 1 != distance[offset] {
          return false;
        }

        buf[offset] = idx as i32;
        true
      })
      .collect::<Vec<bool>>()
      .iter()
      .all(|&x| x)
  }
}
